### PR TITLE
Nucypher Configuration Parsers

### PR DESCRIPTION
**Pull request by KPrasch**
Changes the way file-based configuration is handled for initialising Ursula:

- Deprecates old style config classes
- Introduces .ini config and parsers

This PR is also the beginning of a chain of PRs that has a similar history to PR #370 that aims to builds a command line interface for "spinning-up" Ursula, setting-up node configurations, managing stakes, gleaning network status and executing network simulations.

#377 : Config file and parsers
#378 nucypher-cli prototype and Ursula.stake
#379 CLI + Configuration + Parser Integration
#380 CLI + Blockchain Agency API Integration
#381 Follow-up
#382 Follow-up again

Depends on nucypher #366 (Merged)
Depends on coincurve #23 ofek/coincurve#23 (Merged)
Depends on pyUmbral #200 nucypher/pyUmbral#200 (Merged)

### CODE REVIEW

File: nucypher/config/keys.py [hunk 1]
```diff
@@ -15,7 +16,7 @@
 from web3.auto import w3
 
 from nucypher.config import utils
-from nucypher.config.configs import _DEFAULT_CONFIGURATION_DIR
+from nucypher.config.configs import _DEFAULT_CONFIGURATION_DIR, NucypherConfiguration
```
Comment (cygnusv) [inline nucypher/config/keys.py]: Is `nucypher.config.configs` coming in a later PR? It seems it doesn't exist right now

Comment (KPrasch) [reply nucypher/config/keys.py]: Hmm, ultimately its been deprecated.  Looks like this is cruft from the refactoring.

File: nucypher/characters.py [hunk 2]
```diff
@@ -68,7 +67,7 @@ def __init__(self, is_me=True,
                  crypto_power: CryptoPower = None,
                  crypto_power_ups=None,
                  federated_only=False,
-                 config: CharacterConfiguration = None,
+                 config = None,
```
Comment (tuxxy) [inline nucypher/characters.py]: 👍

File: nucypher/config/utils.py [hunk 3]
```diff
@@ -134,3 +55,165 @@ def check_config_runtime() -> bool:
...
+
+    if config is None:
+        config = configparser.ConfigParser()
+
+    try:
+        config.read(filepath)
+    except:
+        raise  # FIXME
+
+    required_sections = ("provider", "nucypher")
```
Comment (tuxxy) [inline nucypher/config/utils.py]: Neat! 👍

File: nucypher/config/utils.py [hunk 4]
```diff
@@ -134,3 +55,165 @@ def check_config_runtime() -> bool:
...
+    else:
+        raise NotImplementedError
+
+    poa = config.getboolean(section='provider', option='poa', fallback=True)
+    tester = config.getboolean(section='blockchain', option='tester', fallback=False)
+    test_accounts = config.getint(section='blockchain', option='test_accounts', fallback=0)
+    deploy = config.getboolean(section='blockchain', option='deploy', fallback=False)
+    compile = config.getboolean(section='blockchain', option='compile', fallback=False)
+    timeout = config.getint(section='blockchain', option='timeout', fallback=10)
+    tmp_registry = config.getboolean(section='blockchain', option='temporary_registry', fallback=False)
```
Comment (tuxxy) [inline nucypher/config/utils.py]: Also cool 👍

File: nucypher/config/keys.py (hunk 1 above)
Comment (cygnusv) [reply nucypher/config/keys.py]: I checked the top of #382 and it seems that `nucypher/config/configs.py` is still gone there, so it seems `nucypher/config/keys.py` it's outdated?

Comment (KPrasch) [reply nucypher/config/keys.py]: Yes - as mentioned, `configs.py` is ultimately deprecated; however config/keys.py does exist.

Comment (KPrasch) [reply nucypher/config/keys.py]: It looks like the `config.keys.py` module has is not being executed in automated testing! 

Good catch

File: nucypher/config/utils.py [hunk 5]
```diff
@@ -1,104 +1,25 @@
...
-    """Parses a keyfile and returns key metadata as a dict."""
-
-    with open(keypath, 'r') as keyfile:
-        try:
-            key_metadata = json.loads(keyfile)
-        except json.JSONDecodeError:
-            raise NucypherConfiguration.NucypherConfigurationError("Invalid data in keyfile {}".format(keypath))
-        else:
-            return key_metadata
+DEFAULT_CONFIG_DIR = "~"
```
Comment (michwill) [inline nucypher/config/utils.py]: If we don't do it already somewhere else, we probably would want to leave a `# TODO` to leave it somewhere else than `~` (`.local` or whatever)

File: nucypher/config/utils.py [hunk 6]
```diff
@@ -134,3 +55,165 @@ def check_config_runtime() -> bool:
...
+            provider = IPCProvider(config['provider']['ipc_path'])
+        except KeyError:
+            message = "ipc_path must be provided when using an IPC provider"
+            raise Exception(message)  # FIXME
+        else:
+            providers.append(provider)
+    else:
+        raise NotImplementedError
+
+    poa = config.getboolean(section='provider', option='poa', fallback=True)
```
Comment (michwill) [inline nucypher/config/utils.py]: It feels to me that we're writing a schema with datatypes by doing the `getboolean` etc, which kinda suggests that we need some facility for that (hopefully, not XML lol). I think, could be not a bad idea to write the schema in Python.

But I'd leave this to the future: masering configurability of configs is not the focus of what we need right now. Can leave a little `# TODO` note though.

For now, this is great stuff

**Pull request merged**